org-code-javabuilder/playground/src/main/java/org/code/playground/Board.java

```diff
@@ -60,28 +67,77 @@ public void setBackgroundImage(String filename) throws FileNotFoundException {}
    *
    * @param image the image to add. If the image is already on the board, this method does nothing.
    */
-  public void addClickableImage(ClickableImage image) {}
+  public void addClickableImage(ClickableImage image) {
+    if (this.clickableImages.containsKey(image.getId())) {
+      return;
+    }
+    HashMap<String, String> details = this.getImageItemDetails(image);
+    this.clickableImages.put(image.getId(), image);
+    this.playgroundMessageHandler.sendMessage(
+        new PlaygroundMessage(PlaygroundSignalKey.ADD_CLICKABLE_ITEM, details));
+  }
 
   /**
    * Removes the clickable image from the board.
    *
    * @param image the image to remove. If the image is not on the board, this method does nothing.
    */
-  public void removeClickableImage(ClickableImage image) {}
+  public void removeClickableImage(ClickableImage image) {
+    if (!this.clickableImages.containsKey(image.getId())) {
+      return;
+    }
+    this.clickableImages.remove(image.getId());
+    HashMap<String, String> details = this.getRemoveItemDetails(image);
+    this.playgroundMessageHandler.sendMessage(
+        new PlaygroundMessage(PlaygroundSignalKey.REMOVE_ITEM, details));
+  }
+
+  /**
+   * Adds a non-clickable ImageItem to the board.
+   *
+   * @param item the image item to add. If the item is already on the board, this method does
+   *     nothing.
+   */
+  public void addImageItem(ImageItem item) {
+    if (this.items.containsKey(item.getId())) {
+      return;
+    }
+    this.items.put(item.getId(), item);
+    HashMap<String, String> itemDetails = this.getImageItemDetails(item);
+    this.playgroundMessageHandler.sendMessage(
+        new PlaygroundMessage(PlaygroundSignalKey.ADD_IMAGE_ITEM, itemDetails));
+  }
 
   /**
-   * Adds a non-clickable item to the board.
+   * Adds a TextItem to the board.
    *
-   * @param item the item to add. If the item is already on the board, this method does nothing.
+   * @param item the text item to add. If the item is already on the board, this method does
+   *     nothing.
    */
-  public void addItem(Item item) {}
+  public void addTextItem(TextItem item) {
+    if (this.items.containsKey(item.getId())) {
+      return;
+    }
+    this.items.put(item.getId(), item);
+    HashMap<String, String> itemDetails = this.getTextItemDetails(item);
+    this.playgroundMessageHandler.sendMessage(
+        new PlaygroundMessage(PlaygroundSignalKey.ADD_TEXT_ITEM, itemDetails));
+  }
 
   /**
    * Removes the non-clickable item from the board.
    *
    * @param item the item to remove. If the image is not on the board, this method does nothing.
    */
-  public void removeItem(Item item) {}
+  public void removeItem(Item item) {
+    if (!(this.items.containsKey(item.getId()) || this.clickableImages.containsKey(item.getId()))) {
```

Comment:
Why are we checking `this.clickableImages` here as well? We're only removing from `this.items` right?

Reply:
oh good point, I forgot we have `removeClickableImage` 🤦‍♀️